### ISSUE TITLE
Cross compiler for Windows On ARM64?

### ISSUE THREAD

**Issue by ghost** [brechtsanders/winlibs_mingw#203]
Cross-compiling from x64 Windows to ARM64 Windows. What do you think about it?

**Comment by brechtsanders** [comment]
I have been thinking about it for a long time. And not just thinking.

I have actually bought a Windows Dev Kit 2023 to actually have a Windows on ARM system (see #130), and I have tried to get GCC to build on it.

LLVM already works (the version from https://github.com/mstorsjo/llvm-mingw), and I was able to build a lot of packages for native Windows on ARM64 already.

For GCC more work is needed. I asked about this here https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108678 and recently got an answer that GCC 15 stage 1 will have changes specifically for this platform. And the good news is that this stage was just started a few days ago (see: https://gcc.gnu.org/develop.html).

**Comment by kayhayen** [comment]
Just a heads up, I am following this issue in hopes of adding support for it in my Python compiler Nuitka as well, where I use the other arches with great success so far. No pressure, just desiring to not miss out once it happens.

**Comment by brechtsanders** [comment]
As I understood it GCC 15 should be the first one to support the x86_64-w64-aarch64 platform.
Unfortunately I have tried the recent weekly snapshots of GCC with MinGW-w64 and it still won't build for this platform.

I also asked the GCC team what the status is [here](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=108678).

**Comment by Blackhex** [comment]
Hello. We have an experimental native compiler in form of MSYS2 packages available at https://github.com/Windows-on-ARM-Experiments/msys2-woarm64-build. It should be stable for most of the C code; C++ support is in progress but improving rapidly. Please, let us know any feedback, e.g. which form of native toolchain would you prefer if MSYS2 packages are not suitable for some reason.

**Comment by asdetycv1zzc** [comment]
GCC 15.1.0 has been released and it should be able to be compiled targeting aarch64-w64-mingw32
However when I tried to do so, it always drags me into the `undefined reference to Unwind_SJLJ_Exception` trap. 
Is it now really possible to build gcc running on ARM64 Windows devices?

**Comment by towguyhat** [comment]
Any news on this?
Or any way to install msys2 and mingw natively on Windows 10 2021 IoT Ent LTSC 21H1?

**Comment by brechtsanders** [comment]
I'm still not ably to build latest GCC for aarch64-w64-mingw32 either natively or cross-compile from Linux. But I will keep trying...

**Comment by hmartinez82** [comment]
@brechtsanders Yes, I think that GCC support (the so called MINGWARM64 environment in MSYS2) will eventually, some day in the far future, come to existence. But for now my question is: Do you think there's value in creating a WinLibs release for ARM64, but using Clang instead as the default compiler of the toolkit? Akin to MSYS2's CLANGARM64?

**Comment by brechtsanders** [comment]
The LLVM/Clang build against MinGW-w64 is already made available by Martin Storsjö at 
[https://github.com/mstorsjo/llvm-mingw/releases](https://github.com/mstorsjo/llvm-mingw/releases).

Would it be mostly the same as that, or would you miss certain tools I package alongside it?

**Comment by kayhayen** [comment]
Right now, Nuitka is kind of set on using gcc targeting MinGW64 on Windows for all arches, but I understand that GCC project is just not there yet for ARM, not even with version 16.2 yet?

I kind of enjoyed it when winlibs just brought along both and I let my users pick, and lldb and gdb were just both there, and one always worked for me. But I can understand if there's not much point about it. 

Nuitka works around layout differences in Python internal structures with gcc already, and will get back to supporting winlibs with Python3.13+ in the next release. Already works with an experimental option.

As for Clang, I made the choice to force users onto ClangCL effectively for the time being, so as not spend time on another compiler download and setup to maintain and verify for myself.

So I will be waiting and hoping for a gcc.exe for ARM out of winlibs.
